imap_processing/idex/idex_l2a.py

```diff
@@ -273,3 +299,176 @@ def calculate_snr(tof_high: xr.DataArray, hs_time: xr.DataArray) -> NDArray:
     tof_sigma = np.nanstd(baseline_noise, axis=1, ddof=1)
     # Return snr ratio
     return tof_max / tof_sigma
+
+
+def analyze_peaks(
+    tof_high: xr.DataArray,
+    high_sampling_time: xr.DataArray,
+    event_num: int,
+    peaks_2d: list[np.ndarray],
+) -> tuple[NDArray, NDArray]:
+    """
+    Fit an EMG curve to the Time of Flight data around each peak.
+
+    Parameters
+    ----------
+    tof_high : xarray.DataArray
+        The time of flight array.
+    high_sampling_time : xarray.DataArray
+        The high sampling time array.
+    event_num : int
+        Dust event number (for debugging purposes).
+    peaks_2d : list[numpy.ndarray]
+        Nested list of peak indices.
+
+    Returns
+    -------
+    params: numpy.ndarray
+        Array of the EMG fit parameters (mu, sigma, lambda) at the corresponding mu
+        time. Empty slots contain zeros.
+    area_under_emg : numpy.ndarray
+        Array of the area under the EMG curve at that time. Empty slots
+        contain zeros.
+    """
+    # Initialize arrays to store EMG fit results
+    # fit_params: (500, 3) array where the first dimension is the ion flight
+    # time and the second is EMG fit parameters (mu, sigma, lambda) for the peaks at
+    # that flight time.
+    # area_under_emg: (500) array storing the area under each EMG peak at the
+    # corresponding flight time.
+    # The Ion flight time can be used to get the estimated mass from the mass_scale
+    # array.
+    fit_params = np.zeros((500, 3))
+    area_under_emg = np.zeros(500)
+    for peak in peaks_2d[event_num]:
+        # Take a slice of 5 samples on either side of the peak
+        start = max(0, peak - 5)
+        end = min(len(tof_high), peak + 6)
+
+        time_slice = np.asarray(high_sampling_time[start:end].data)
+        tof_slice = np.asarray(tof_high[start:end].data)
+
+        param = fit_emg(time_slice, tof_slice, event_num)
+        if param is not None:
+            area = calculate_area_under_emg(time_slice, param)
+            # Center of fitted gaussian (time)
+            emg_mu = param[0]
+            # Round calculated time to the nearest int
+            # If that index is already taken, keep increasing the index by one
+            # until we find an empty slot.
+            # This ensures we don't overwrite existing data when we have multiple peaks
+            # close to the same flight time.
+            if emg_mu < 0:
+                logger.warning(
+                    f"Warning: The EMG fit resulted in a negative mu "
+                    f"(center of gaussian) value: {emg_mu}."
+                )
+
+            idx = max(0, round(emg_mu))
+            while np.all(fit_params[idx:] != 0) and idx < 500:
+                idx += 1
```

Comment:
I think you can find the next nonzero value with numpy rather than the loop. Something like this, (it is probably not exactly right)
`idx = np.nonzero(np.all(fit_params != 0, axis=-1))`